### ISSUE TITLE
AzCopy 10.13.0 is ten times slower than 10.3.4

### ISSUE THREAD

**Issue by dmitrychastikov** [Azure/azure-storage-azcopy#1620]
### Which version of the AzCopy was used? 
10.13.0

### Which platform are you using? (ex: Windows, Mac, Linux)
Windows

### What command did you run?
 copy https://stgnblhns74ensq2qmq.blob.core.windows.net/meridianstreams/?se=2021-11-16t20%3A17%3A55z&sig=-REDACTED-&sp=rl&spr=https&srt=sco&ss=b&st=2021-11-16t12%3A17%3A55z&sv=2018-03-28 https://stgcloudbackup.blob.core.windows.net/2021-11-14-stgnblhns74ensq2qmq?se=2021-11-16t20%3A17%3A54z&sig=-REDACTED-&sp=racwdl&spr=https&srt=sco&ss=b&st=2021-11-16t12%3A17%3A54z&sv=2018-03-28 --recursive=true --overwrite ifSourceNewer 

### What problem was encountered?
In total enumeration of 68000 blobs of the source and comparison with 68000 of the destination with copying 1 or 2 blobs takes 1.5 minutes with AzCopy 10.3.4. The same job using AzCopy 10.13.0 in reality takes closer to 30minutes. It reports 
```
Job 55d03ee1-98c8-6d47-7e4c-8e87f81a1099 summary
Elapsed Time (Minutes): 17.5001
Number of File Transfers: 67993
Number of Folder Property Transfers: 0
Total Number of Transfers: 67993
Number of Transfers Completed: 1
Number of Transfers Failed: 0
Number of Transfers Skipped: 67992
TotalBytesTransferred: 317628
Final Job Status: CompletedWithSkipped
``` 
but 10 minutes of scanning takes place before.

Even 17 minutes against 1.5 is a huge performance degradation.

And how it works with AzCopy 10.3.4:
```
Job 7acd6f00-d61b-2f43-69ea-f8285eed0876 summary
Elapsed Time (Minutes): 1.6246
Total Number Of Transfers: 67993
Number of Transfers Completed: 2
Number of Transfers Failed: 0
Number of Transfers Skipped: 67991
TotalBytesTransferred: 635256
Final Job Status: CompletedWithSkipped

```
I did this on a b2ms Azure VM, the same we do nightly backups for several million blobs. On Sundays full copy is created and 10.3.4 could do that. Now 10.13.0 failed (one night is not long enough) basically to enumerate them. Incremental copy is usually just a tiny fraction of the total. Several attempts using 10.13.0 and 10.12.2 and 10.3.4 showed that only 10.3.4 can be used.
   
### How can we reproduce the problem in the simplest way?
The simplest is to copy one storage (container) to another one, and then do it again and again with different versions.

### Have you found a mitigation/solution?
Solution appeared to be to use 10.3.4 that has been working for ~18months.

**Comment by ASGgregl** [comment]
I'm having the same issue, except that I was still getting good performance using version 10.10.0, and started seeing the problem when I upgraded to 10.12.2.  Problem continues with 10.13.0.  I am using the sync command.  Sample:

azcopy sync "https://REDACTED.blob.core.windows.net/access-products/FIXES?REDACTED" "https://REDACTED.blob.core.windows.net/access-products"/FIXES?REDACTED" --recursive=true --delete-destination=true

**Comment by zezha-msft** [comment]
Thanks @dmitrychastikov  @ASGgregl for reaching out!

Have you tried setting the following to true with 10.13?

INFO: Name: AZCOPY_DISABLE_HIERARCHICAL_SCAN
Current Value:
Description: Applies only when Azure Blobs is the source. Concurrent scanning is faster but employs the hierarchical listing API, which can result in more IOs/cost. Specify 'true' to sacrifice performance but save on cost.

My guess is that your container structure contains a lot of folders with very few files in them, is this correct by any chance?

**Comment by dmitrychastikov** [comment]
>>> My guess is that your container structure contains a lot of folders with very few files in them, is this correct by any chance?
It is exactly the other way around, just a lot of blobs.

>>> Specify 'true' to sacrifice performance but save on cost.

I tell you that the performance is bad now. 10 times worse than it used to be in previous versions. And your proposition is to sacrifice PERFORMANCE even more???

So far we are stuck to 10.3.4 that is reasonably fast. Although the amount of data blobs is growing and even with that version we have to optimize the processes to fit the full backups into one night.

**Comment by ASGgregl** [comment]
I am having the slowdown with both Azure blob and local files as the source.  I notice that the flag you suggested "_Applies only when Azure Blobs is the source._ " so it won't really help me.

**Comment by zezha-msft** [comment]
@dmitrychastikov no, I believe the suggestion of AZCOPY_DISABLE_HIERARCHICAL_SCAN is a reasonable thing to try here. 

Since you mentioned that the scanning in the more recent versions seems to be taking a lot longer than 10.3.4 for your dataset, so I thought of one relevant change that happened in 10.5:

1. Improved scanning performance for most cases by adding support for parallel local and Blob enumeration.

Parallel scanning here means going into the virtual directories using the hierarchical listing API to try to speed up the overall scanning performance. However, there are edge cases where it may be slower than just using the flat listing API in a single-threaded fashion, e.g. when there are lots of folders with only very few blobs in them.

But it's interesting that your dataset is not the edge case that we know of, could you please describe the structure a bit more? Are most of your blobs just under the container directly? And where is your VM running from? Is it closer to the source region or the destination region?

Thanks.

**Comment by zezha-msft** [comment]
@ASGgregl ah I see, were you using sync though? If yes then it may come into the picture as AzCopy needs to scan both the source and the destination in that case.

Do you mind giving AZCOPY_DISABLE_HIERARCHICAL_SCAN a try? 

I want to reassure you and @dmitrychastikov that we'll take this problem seriously and work with you to get to the bottom of the apparent performance gap.

**Comment by dmitrychastikov** [comment]
Our structure is pretty simple and cannot be an edge case, just a lot of containers with a lot of blobs inside distributed pretty much equally in virtual folders Level1/Level2/blob having just a few level2 folders and blobs inside many level1.
We copy always within the same region WEU->WEU, EUS->EUS; storage accounts are geo-redundant RA-GRS if that matters.

Here is the command that is used. Also may be important that we use concurrency settings that allowed to squeeze 1-million-blobs-account into an hour of copying. This was figured out two years ago and works good so far with 10.3.4. Keep in mind that during a few hours window we have to copy several 1-million-blobs-accounts (and as of today some of them are bigger than 1.6M)
   $env:AZCOPY_CONCURRENCY_VALUE = "256"
    & c:\backup\tools\$azcopyVersion\azcopy copy "$SourceStorage/$SASSRC" "$Destination$SASDEST" --recursive=true --overwrite ifSourceNewer  >>$logFile

Just use a test storage account with a reasonably big set of blobs, I used ~65000 blobs.
Copy into empty destination using 10.3.4
Copy into another empty destination using 10.13.0
Compare times.
Copy into filled in destination using 10.3.4
Copy into filled in destination using 10.13.0
Compare times.
My comparison shows ~10 times difference. What showed yours?

**Comment by ASGgregl** [comment]
I tested 10.10 vs 10.13 with AZCOPY_DISABLE_HIERARCHICAL_SCAN set to both true and false.  I did not see a significant difference with AZCOPY_DISABLE_HIERARCHICAL_SCAN set to true or false.  10.10 consistently did the sync in the 4-4:30 range and 10.13 was between 8 and 9 minutes.  This was with local files as the source.

**Comment by fraenkDUS** [comment]
Similar issue here, version 10.17 is awfully slow not even 10Mbit/s, but 10.10 gives over 100Mbit/s in download.
The job is to download one folder containing 5-10 million small files around 25KB each from Azure Files (StorageV2 (general purpose v2, hot, large file share, transaction optimized).

Tried to disable any performance decreasing option in version 10.17 with copy and sync while increasing concurrency to 2048, but no luck. Downgrading to 10.10 did the trick but it doesn't recognize previous jobs...

**Comment by dphulkar-msft** [comment]
We recommend trying the latest version of Azcopy (10.28.1) to see if it resolves the issue. We will go ahead and close the ticket for now. Please feel free to reopen it if you continue to experience the issue after updating to the latest version.

Thank you for your patience!
